docs: complete config and grc docu (#1898)
GIScience/openrouteservice: docs/run-instance/configuration/how-to-set-properties.md

* [Different Notations](#different-notations) for the same configuration properties
* [Internal defaults](#internal-defaults): Where are those defined, how to view the internal default values
* [Different ways to set individual Values](#different-ways-to-set-individual-values)
    * Place a [Default Config File](#default-config-file) named `ors-config.yml` in one of the directories where
      openrouteservice looks for it
    * or specify a [Custom Config File](#custom-config-file)
    * override properties with [Environment variables](#environment-variables) in the shell or in an environment file
    * override properties with [JVM Program Arguments](#jvm-program-arguments)
    * override properties with [Spring Run Arguments](#spring-run-arguments)
    * [**Priority of these different property sources**](#priority-of-these-different-property-sources)
* [Docker Scenarios](#docker-scenarios)
* [Defining defaults for all routing profiles](#defining-defaults-for-all-routing-profiles) in
  `ors.engine.profile_default`
* [Graph build properties loaded from graph directories](#graph-build-properties-loaded-from-graph-directories)

## Different Notations

> The configuration or openrouteservice can be specified in different notations.

Reviewer: ```suggestion
The configuration of openrouteservice can be specified in different notations.
```

Which format to best use in which scenario is documented in the config documentations for
[JAR](../running-jar.md), [WAR](../running-war.md) and [Docker](../running-with-docker.md).

* `.yml` is the default configuration format since version 8. You can find
  an [example configuration file](https://github.com/GIScience/openrouteservice/blob/main/ors-config.yml) with all
  available configuration options. Only a minimal set of properties is active, all others are commented out.
* `.env` files for Docker setup. There is also
  an [example env file](https://github.com/GIScience/openrouteservice/blob/main/ors-config.env) that you can download
  and customize.
* `.json` config file: In the past openrouteservice was configured [via JSON file](json.md). This configuration method

>   has been **deprecated** and will be eventually removed. Therefore, we strongly discourage you from using it. If you

Reviewer: ```suggestion
  has been **deprecated** and removed in ORS version 9. Therefore, we strongly discourage you from using it. If you
```

  have an old JSON config, please consider
  to [migrate to the new config](https://github.com/GIScience/ors-config-migration#usage).

All of the above described config files can contain the same logic application properties.

For example, the property `ors.engine.profiles.car.enabled` would look like this

in `*.yml`

```yaml
ors:
    engine:
        profiles:
            driving-car:
                enabled: true
```

in `*.env` one of

```shell
ors.engine.profiles.driving-car.enabled=true
ORS_ENGINE_PROFILES_DRIVING_CAR_ENABLED=true
```

as Spring or JVM run parameter

```shell 
ors.engine.profiles.driving-car.enabled=true
```

::: tip
In this documentation we mostly use the dot notation to name a configuration property
because it is more compact than e.g. YAML style with line breaks and indentation
and better readable than uppercase-underscore environment variable style.
:::

## Internal Defaults

openrouteservice has sensible defaults and can be started with a minimal configuration,

> see [Configuration Properties](/run-instance/configuration/configuration-properties.md#minimal-configuration).

Reviewer: ```suggestion
see [Configuration Properties](configuration-properties.md#minimal-configuration).
```

The internal default values are defined in the file
[
`ors-api/src/main/resources/application.yml`](https://github.com/GIScience/openrouteservice/blob/main/ors-api/src/main/resources/application.yml).

Besides default values for most configuration properties,
there are also **default routing profiles** for all supported modalities with sensible defaults for each of them.
These default profiles have the names of their flag encoders.

> For more details see [profiles](/run-instance/configuration/engine/profiles/index.md).

Reviewer: ```suggestion
For more details see [profiles](engine/profiles/index.md).
```

The openrouteservice [releases](https://github.com/GIScience/openrouteservice/releases) also contain downloadable
example `ors-config.yml` and `ors-config.env` files as release assets,
where all properties with their defaults are included as commented lines.

> ### Configuration defaults output

Reviewer: Is this feature tested?

Author: probably not...

      java -jar ors.jar
      ```

If both are specified, the program argument wins.
When a custom config file is used, `ors-config.yml` files are ignored.

### Environment Variables

All or single configuration parameters can be overridden/defined as environment variables.

```shell 
export ors.engine.profile_default.build.source_file=./osm_file.pbf 
export ors.engine.profiles.driving-car.enabled=true
java -jar ors.jar 
```

This is especially useful in contexts where you want to run
openrouteservice in containers such as with [docker](../running-with-docker.md).

Every property also corresponds to an environment variable name in *uppercase letters* and with *underscores* replacing
*dots*, so e.g.

- `ORS_ENGINE_PROFILE_DEFAULT_BUILD_SOURCE_FILE` replaces `ors.engine.profile_default.build.source_file`
- `ORS_ENGINE_PROFILES_CAR_ENABLED` replaces `ors.engine.profiles.car.enabled`

Consequently, the following commands are equivalent to the last example above:

```shell
  export ORS_ENGINE_SOURCE_FILE=./osm_file.pbf
  export ORS_ENGINE_PROFILES_CAR_ENABLED=true
  java -jar ors.jar
```

Note, that there are different ways to define environment variables,

> e.g. in [docker compose files](/run-instance/running-with-docker#set-openrouteservice-properties-in-docker-compose-yml)

Reviewer: ```suggestion
e.g. in [docker compose files](../running-with-docker#set-openrouteservice-properties-in-docker-composeyml),
```

Author: you suggested change breaks the link

or

> [env files](/run-instance/running-with-docker#set-openrouteservice-properties-in-an-environment-file)

Reviewer: ```suggestoin
[env files](../running-with-docker#set-openrouteservice-properties-in-an-environment-file)
```

Author: don't like relative upwards links

> or in `setenv.sh` files in [tomcat setups](/run-instance/running-war#configure).

Reviewer: ```suggestion
or in `setenv.sh` files in [tomcat setups](../running-war#configure).
```

Author: see above

### JVM Program Arguments

All or single configuration parameters can be overridden/defined as JVM run arguments:

```shell 
java -jar -Dors.engine.profile_default.build.source_file=./osm_file.pbf -Dors.engine.profiles.driving-car.enabled=true ors.jar 
```

### Spring Run Arguments

All or single configuration parameters can be overridden/defined as Spring run arguments:

```shell 
java -jar ors.jar --ors.engine.profile_default.build.source_file=./osm_file.pbf --ors.engine.profiles.driving-car.enabled=true
```

### Priority of these different property sources

If properties are set in several configuration sources,
the priority is as follows:

* Spring Run Arguments override
* JVM Program Arguments override
* Environment Variables override
* Properties defined in the single used config file override

> * Internal defaults

Reviewer: ```suggestion
If properties are set in several configuration sources,
the priority is as follows. 1 (highest priority) to 5 (lowest):

1. Spring Run Arguments override
2. JVM Program Arguments override
3. Environment Variables override
4. Properties defined in the single used config file override
5. Internal defaults
```

The following image illustrates, how a config yml overrides some of the internal defaults.
The merge result is shown in the middle.
The red lines are examples for properties defined on both sides - the user config wins.
The commented lines (green) are just included for better readability.

![](config-overwrite/config-overwrite-merge-files.drawio.png )

## Docker Scenarios

> A openrouteservice running in a container works in the same way as described in this chapter.

Reviewer: ```suggestion
An openrouteservice instance running in a container works in the same way as described in this chapter.
```

The difference is, that the openrouteservice has a docker container as its environment,

> with a separat file system (different paths) and individual environmen variables.

Reviewer: ```suggestion
with a separate file system (different paths) and individual environment variables.
```

The lookup for configuration files is always related to the internal environment,
this has to be respected when configuring paths in the YAML config or

> when specifying the config file in the parameter ors_config_location.

Reviewer: ```suggestion
when specifying the config file in the parameter `ors_config_location`.
```

Like in a "normal" setup, all configuration properties can be set as environment variables.
This is a more common way for configuring applications in docker containers,

> in contrast to applications running directly on the host sytem,

Reviewer: ```suggestions
in contrast to applications running directly on the host system,
```

because in the docker container there is a fresh and independent environment.

> [Running with Docker](/run-instance/running-with-docker#configure) describes in detail,

Reviewer: ```suggestion
[Running with Docker](../running-with-docker#configure) describes in detail,
```

Author: see above

One side effect is, that properties of specific profiles, that are defined in the internal defaults,
cannot be overridden by properties in the user's `profile_default`!
In the example above the user has defined `ors.engine.profile_default.maximum_snapping_radius: 300`
but the profile foot-walking has an internal default `400` for this property.
The internal value for the specific profile wins over the user's `profile_default`:

![](config-overwrite/config-overwrite-merge-defaults.drawio.png)

## Graph build properties loaded from graph directories

The configured profile properties in `ors.engine.profiles.<profile>.build` (and `ors.engine.profile_default.build`)
are used to _build_ a graph.
openrouteservice stores these build properties in the file `graphs/<profile>/graph_info.yml` when a graph is computed.

When openrouteservice is _loading_ an existing graph instead of building it,
the persisted build parameters are again loaded from this file,
locally configured build parameters for the affected routing profile are overridden.

Therefore, the file `graph_info.yml` must not be edited!
The content of this file is also used for repository lookups by

> the ([Graph Repo Client](/technical-details/graph-repo-client/index.md)). 

Reviewer: ```suggestion
the ([Graph Repo Client](../../technical-details/graph-repo-client/index.md)). 
```
